### DISCUSSION TITLE
Issue converting PFM to TIF across ImageMagick versions

### DISCUSSION THREAD

**Discussion by DavidBorts** [ImageMagick/ImageMagick#6590] [Help]
### ImageMagick version

7.1.1-12

### Operating system, version and so on

Windows 11, personal PC

### Description

Hey, all.

I am a novice ImageMagick user running into some unexpected behavior when converting PFM images to TIF. 

On my own machine (version 7.1.1-12), I am reliably able to perform this conversion, with a command like:
`magick ./img.pfm ./img.tif`

However, a project requires me to perform this same conversion on an HPC cluster. They have their own, older build of ImageMagick - one that's configured a little differently (since they're using redhat). They're using 6.9.12-86:
![image](https://github.com/ImageMagick/ImageMagick/assets/78828451/31fab8a1-8d60-4821-9240-fe997e66ae74)

On the cluster, I would typically perform the conversion by running:
`convert ./img.pfm ./img.tif`

However, the TIFF image produced on the cluster is totally off:
![image](https://github.com/ImageMagick/ImageMagick/assets/78828451/26db9856-3c0c-422b-a416-8337c5bb08a8)

Looking into the individual pixel values, it becomes clear that something is wrong:
![image](https://github.com/ImageMagick/ImageMagick/assets/78828451/656b3c61-3d7c-4b18-9a35-cd93370c2b75)
These values are enormous!

Unfortunately, though, this doesn't just seem to be an issue of normalization. When I normalize the above image, and subtract it from the correctly-converted tif (made with my newer build of ImageMagick), I find that there is still a slight difference (<1e-4 per pixel). Here's what the difference looks like visualized:
![image](https://github.com/ImageMagick/ImageMagick/assets/78828451/0249539e-fc53-4b00-8771-5b08bf2eedd9)

My question is: what might be causing these different outputs? Did something change in the convert utility between these two versions? Is there some flag or option that I should be passing into the older version, that is set to true or false by default on newer builds? I'm afraid I'm not familiar enough with ImageMagick to have any intuition as to what's going wrong. 

Any help would be much appreciated - thanks!

**Comment by urban-warrior** [comment]
Post a link to your source image.  We need to reproduce the issue before we can offer advice.

**Reply by DavidBorts** [reply]
Apologies - here is a link to a drive folder with 2 images that can be used to reproduce the issue.

https://drive.google.com/drive/folders/1tRspRDbH4HuxGyoY1Ura8OVAVO58fLVB?usp=sharing

**Comment by urban-warrior** [comment]
We tried your image with ImageMagick 6.9.12-93, the current release, and the image rendered properly.  Its possible there was a bug fix between the -86 and the -93 release of 6.9.12.  Try converting the image to another format with -86.  Does it look okay?  If so, the issue is with the TIFF coder.  If its still errant, the issue is with the PFM coder.

**Reply by DavidBorts** [reply]
Interesting... I'll try this asap and get back to you

**Reply by DavidBorts** [reply]
I just tried converting to PNG instead, both on the cluster and on my local machine - I got the exact same image from both, which appears to render correctly.

**Reply by DavidBorts** [reply]
Another update - I tried (on the cluster) first converting to .hdr, and then to .tif - the resulting image rendered correctly, but in terms of pixel data, is still different, even when normalized.

**Comment by snibgo** [comment]
The images convert fine for me, with:

6.9.9-40 Q16 (non-HDRI)
6.9.12-72 Q16-HDRI
7.1.1-15 Q16-HDRI

The images are grayscale, but seem to be Bayer images, ie each pixel represents one channel, either red, green or blue.

**Reply by DavidBorts** [reply]
This is true -the two images in the folder are the same bayer mosaic before and after a white balance operation

What was the exact command that you ran on the 6.9.* builds? Perhaps my mistake lies therein.

**Comment by fmw42** [comment]
try

`convert image.pfm -define quantum:format=floating-point -depth 16 image.tiff`

or

`convert image.pfm -define quantum:format=floating-point -depth 32 image.tiff`

**Reply by DavidBorts** [reply]
I'll give this a try and report back!

**Comment by snibgo** [comment]
> What was the exact command that you ran on the 6.9.* builds...

I did nothing fancy. On v7:
```
magick imagemagick\test_img_1.pfm %TEMP%\t3.tiff
```
On v6:
```
convert imagemagick\test_img_1.pfm %TEMP%\t2.tiff
```

`magick -list format |grep -i tiff` (or `convert ...`) say:
```
(LIBTIFF, Version 4.3.0)
(LIBTIFF, Version 4.4.0)
(LIBTIFF, Version 4.5.1)
```
I suggest you try writing to other formats, such as PNG and JPG. Are they okay?

I also suggest converting other images to TIFF. Do they work okay?

Your result seems mostly black or white with very few grays, with negation or wraparound of tones. It is also smaller (1229x922 instead of 4048x3036), and PNG instead of TIFF. You might check the process that made that file.

**Comment by snibgo** [comment]
I doubt that "`-define quantum:format=floating-point -depth 32`" will make any difference. IM notices that the input is floating-point, so the output will also be floating point.

```
C:\web\im>%IMG7%magick %TEMP%\t2.tiff info:
D:\temp\t2.tiff TIFF 4048x3036 4048x3036+0+0 32-bit Grayscale Gray 46.8822MiB 0.000u 0:00.000
```

Of course, some software mis-reads floating-point TIFFs. Hence my comment: how did you make the PNG from the TIFF? You might also try "`-define quantum:format=unsigned -depth 16`".
